org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/Main.java

```diff
@@ -1741,13 +1742,38 @@ private boolean checkVMVersion(long minimalSupportedVersion) {
 	}
 	return ClassFileConstants.getComplianceLevelForJavaVersion(majorVersion) >=minimalSupportedVersion;
 }
+
+protected SourceInfo getSourceInfo(int index) {
+	if (this.sourceInfos == null) {
+		return new SourceInfo(index, getOrNull(this.filenames, index), getOrNull(this.encodings, index), getOrNull(this.modNames, index), getOrNull(this.destinationPaths, index));
+	}
+	return this.sourceInfos[index];
+}
+
+private String getOrNull(String[] arr, int i) {
+	if (arr == null) {
+		return null;
+	}
+	return arr[i];
+}
+
 /*
  *  Low-level API performing the actual compilation
  */
 public boolean compile(String[] argv) {
 	// decode command line arguments
 	try {
 		configure(argv);
+		// sort array by file names so we have a consistent order of compiling / handling them
```

Comment:
this might even be pushed to to perform compile but as the EclipseCompiler extending this class uses some own indexed data structure this need to be carefully done in another PR...